File: python/sorting_algorithms/algos/comb_sort.py
from typing import Sequence, Optional
from ..base import _setup_tracer, _finalize, SortTracer, Number


def comb_sort(data: Sequence[Number], *, trace: bool = False, tracer: Optional[SortTracer] = None, shrink: float = 1.3):
    arr = list(data)
    tracer, ret_trace = _setup_tracer("comb_sort", trace, tracer)
    gap = len(arr)
    swapped = True
    while gap > 1 or swapped:
        gap = max(1, int(gap / shrink))
        swapped = False
        for i in range(len(arr) - gap):
            j = i + gap
            if tracer:
                tracer.record(arr, (i, j), f"gap={gap}")
            if arr[i] > arr[j]:
                arr[i], arr[j] = arr[j], arr[i]
                swapped = True
                if tracer:
                    tracer.record(arr, (i, j), "swap")
    return _finalize(arr, tracer, ret_trace)

File: c/src/comb_sort.c
#include "../sorting_algorithms.h"

void comb_sort(int *arr, size_t n) {
    double shrink = 1.3;
    size_t gap = n;
    int swapped = 1;
    while (gap > 1 || swapped) {
        gap = (size_t)(gap / shrink);
        if (gap < 1) gap = 1;
        swapped = 0;
        for (size_t i = 0; i + gap < n; ++i) {
            size_t j = i + gap;
            if (arr[i] > arr[j]) {
                int tmp = arr[i]; arr[i] = arr[j]; arr[j] = tmp;
                swapped = 1;
            }
        }
    }
}
